framework/src/physics/DiffusionCG.C

```diff
@@ -0,0 +1,231 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "DiffusionCG.h"
+#include "MooseVariableBase.h"
+
+// Register the actions for the objects actually used
+registerMooseAction("MooseApp", DiffusionCG, "add_kernel");
+registerMooseAction("MooseApp", DiffusionCG, "add_bc");
+registerMooseAction("MooseApp", DiffusionCG, "add_variable");
+registerMooseAction("MooseApp", DiffusionCG, "add_preconditioning");
```

Comment:
it would be good to inherit this one from the base class. Unfortunately this is not how it works.
We cannot just register the task to `[Physics]` as the parameters for the preconditioning are nested under `Physics/DiffusionCG/some_name`. 
It does not seem to me that I can register to DiffusionPhysicsBase either. 

Maybe the way forward is to add the task dynamically at construction?

Reply:
I think there are cases where an action inherits from another where you don't want to automatically inherit its associated tasks. I believe the `add_bounds` task is one I reviewed of yours today where you were deliberately trying to disambiguate bounds from kernel/auxkernel but you inherit the `AddBoundsAction` from `AddKernelAction`.

You could define a macro in the base class
```c++
#define registerDiffusionPhysicsBaseTasks(derived_name) \
  registerMooseAction("MooseApp", derived_name, "add_preconditioning")
```

Reply:
That works. I ll add that

Reply:
I was trying to understand this afternoon why the coverage is relatively low despite your framework `Physics` class and testing. It is tied to the task registration. If a developer implements one of your `virtual` methods without registering the task then they will likely be surprised when their objects don't get added. So maybe we do want tasks registered in a base action to, by default, propagate to the derived class. And then we could offer a way to de-register an inherited task if desired. What do you think? If you're in agreement, do you want to do that here in this PR or we can create an issue?

Reply:
So essentially I think all the tasks that you check in `PhysicsBase::act` should be associated tasks for derived classes. This feels right given your `virtual` interfaces; e.g. a developer who implements any of your `virtual` methods will get the behavior they might expect without having to remember to register a task or call a macro from the base class. This would also have a side effect of making the coverage in `PhysicsBase::act` 100%

Reply:
> This would also have a side effect of making the coverage in PhysicsBase::act 100%

actually since I would de-register DiffusionCG and DiffusionFV from all the tasks they do not use, and there are no other Physics in the framework, I would not get any better coverage from this.

Reply:
I do agree that this is how the `virtual` interfaces should work.

Reply:
> actually since I would de-register DiffusionCG and DiffusionFV from all the tasks they do not use

Why would you do this? You have empty implementations for the virtual methods in the base class so why should a user need to bother de-registering tasks? It seems like a user should only have to think about implementing the virtual methods and that's that

Reply:
looks better in the debug log. You dont see plenty of tasks being triggered doing nothing

Reply:
Well I still think there should be registration of tasks in `PhysicsBase` and that they should be inherited. The macro is a workaround, but I think it's a poor one. We need to make sure that tasks like your `init_physics` task should run for all physics. Right now someone who calls your `dimension()` API and relies on the result will get undefined behavior because your `_dim` member is uninitialized

Reply:
I agree with that. 
Just letting you know I plan to deregister as well

Reply:
fair enough

Reply:
Thinking more about this. 
I m thinking register everything remotely available / needed (all the tasks for the add...) in PhysicsBase
then if the relevant routine is not implemented, deregister automatically.

this would require making act() final too.
and I dont think "check if a routine is implemented and not just {} " is possible in C++ from my googling

back to thinking

Reply:
From the registerMooseAction macro we can see there's a lot of meta data that we are not going to be able to get by simply adding a call in the parent class that would work for all the derived classes. 
```
      Registry::addAction<classname>({app, #classname, "", task, __FILE__, __LINE__, "", ""})
```
classnames, file and line are not something the parent class knows. Even the app may not be the same app.
Then file and line are only for debugging so that s something I can overlook for now.

That data is later processed (into the action warehouse I think) into a list of tasks per action.
I think if we did something, we d work in the action warehouse directly instead of the registry because of the missing data. Still seems tricky to have calls in the base action that register tasks for derived actions.

So Instead, let's allow ourselves to have the calls registering the tasks in the derived actions. A few options:
- listing every task every time
Not super clean. Especially since act() is no longer defined in the derived class.

- the macro you suggested
The macro does make it a one-liner, we are still adding additional tasks on top of it

- looping over base classes of the physics to register the tasks of the base classes
not possible in C++ based on my googling. You d have to list them explicitly

I think I want to have option 2 + the following changes
- make the PhysicsBase addXXX return a warning if they are called. It means too many tasks got registered
- keep track of the required tasks for a physics class, and have a check that all the required tasks are registered for the derived classes.

Reply:
@loganharbour @dschwen 

If there is a better way to 'inherit' these tasks from the parent physics class.
registerMooseAction is necessarily called from the derived class. 
looping over all actions and dynamic_casting is somewhat of an option but not my favorite as it would mean registering tasks to the virtual parent classes (and checking all the derived ones against each of their parent)